### PR TITLE
Convert DomainCreator to use factory map

### PR DESCRIPTION
**Pull request by wthrowe**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Elliptic/Executables/Elasticity/SolveElasticity.hpp
Comment (nilsvu) [inline src/Elliptic/Executables/Elasticity/SolveElasticity.hpp:9]: Is the registration actually necessary for the domain creators / do you know why it is? Since the domain creators are not PUPable it seems to me that Charm doesn't need to know about the subclasses.

**Review by nilsvu — DISMISSED** [review]
LGTM

File: src/Elliptic/Executables/Elasticity/SolveElasticity.hpp
Comment (wthrowe) [reply src/Elliptic/Executables/Elasticity/SolveElasticity.hpp:9]: This actually registers the maps, not the creators.  I agree that the name is somewhat misleading.

Comment (nilsvu) [reply src/Elliptic/Executables/Elasticity/SolveElasticity.hpp:9]: Oh I see, fine then

File: src/Domain/Creators/DomainCreator.hpp
Comment (kidder) [inline src/Domain/Creators/DomainCreator.hpp:16]: Is this include really needed?  Also not your fault, but the includes and forward declarations for this file are completely messed up.  Can you fix them while touching this file?  If it turns out to be a mess because of missing includes in files that included this one, then maybe just file an issue

File: src/Domain/Creators/Factory1D.hpp
Comment (kidder) [inline src/Domain/Creators/Factory1D.hpp:17]: `#include "Utilities/TMPL.hpp"`

File: src/Domain/Creators/Factory2D.hpp
Comment (kidder) [inline src/Domain/Creators/Factory2D.hpp:18]: `#include "Utilities/TMPL.hpp"`

File: src/Domain/Creators/Factory3D.hpp
Comment (kidder) [inline src/Domain/Creators/Factory3D.hpp:23]: `#include "Utilities/TMPL.hpp"`

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash if you wish

File: src/Domain/Creators/DomainCreator.hpp
Comment (wthrowe) [reply src/Domain/Creators/DomainCreator.hpp:16]: I believe it is needed to destroy the temporary returned from `functions_of_time()`.  (That function is where I got errors, in any case.)

I'll take a look at the other includes.

Comment (kidder) [reply src/Domain/Creators/DomainCreator.hpp:16]: ah yes, that makes sense

**Comment by wthrowe** [comment]
I've added another commit fixing the includes in DomainCreator.hpp.  Luckily, nothing else was relying on them.

**Review by kidder — DISMISSED** [review]

**Review by nilsvu — DISMISSED** [review]
No need to wait for my re-approval after this gets rebased

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@wthrowe this needs a rebase to pass CI

**Review by nilsvu — APPROVED** [review]

**Pull request merged**
